packages/contract-wrappers/src/contract_wrappers/coordinator_wrapper.ts

```diff
@@ -0,0 +1,672 @@
+     * @param   orderTransactionOpts  Optional arguments this method accepts.
+     * @return  Transaction hash.
+     */
+    @decorators.asyncZeroExErrorHandler
+    public async batchHardCancelOrdersAsync(
+        orders: SignedOrder[],
+        orderTransactionOpts: OrderTransactionOpts = { shouldValidate: true },
+    ): Promise<string> {
+        assert.doesConformToSchema('orders', orders, schemas.ordersSchema);
+        const makerAddress = getMakerAddressOrThrow(orders);
+        assert.doesConformToSchema('orderTransactionOpts', orderTransactionOpts, orderTxOptsSchema, [txOptsSchema]);
+        await assert.isSenderAddressAsync('makerAddress', makerAddress, this._web3Wrapper);
+
+        const data = this._transactionEncoder.batchCancelOrdersTx(orders);
+        const transaction = await this._toSignedZeroExTransactionAsync(data, makerAddress);
+        return this._sendTransactionAsync(
+            transaction,
+            makerAddress,
+            transaction.signature,
+            [],
+            [],
+            orderTransactionOpts,
+        );
+    }
+
+    /**
+     * Hard cancellation on Exchange contract.
+     * Cancels all orders created by makerAddress with a salt less than or equal to the targetOrderEpoch
+     * and senderAddress equal to msg.sender (or null address if msg.sender == makerAddress).
+     * @param   targetOrderEpoch             Target order epoch.
+     * @param   senderAddress                Address that should send the transaction.
+     * @param   orderTransactionOpts         Optional arguments this method accepts.
+     * @return  Transaction hash.
+     */
+    @decorators.asyncZeroExErrorHandler
+    public async hardCancelOrdersUpToAsync(
+        targetOrderEpoch: BigNumber,
+        senderAddress: string,
+        orderTransactionOpts: OrderTransactionOpts = { shouldValidate: true },
+    ): Promise<string> {
+        assert.isBigNumber('targetOrderEpoch', targetOrderEpoch);
+        assert.doesConformToSchema('orderTransactionOpts', orderTransactionOpts, orderTxOptsSchema, [txOptsSchema]);
+        await assert.isSenderAddressAsync('senderAddress', senderAddress, this._web3Wrapper);
+        return this._exchangeInstance.cancelOrdersUpTo.sendTransactionAsync(targetOrderEpoch, {
+            from: senderAddress,
+            gas: orderTransactionOpts.gasLimit,
+            gasPrice: orderTransactionOpts.gasPrice,
+            nonce: orderTransactionOpts.nonce,
+        });
+        // // todo: find out why this doesn't work (xianny)
+        // const data = this._transactionEncoder.cancelOrdersUpToTx(targetOrderEpoch);
+        // const transaction = await this._toSignedZeroExTransactionAsync(data, senderAddress);
+        // return this._sendTransactionAsync(
+        //     transaction,
+        //     senderAddress,
+        //     transaction.signature,
+        //     [],
+        //     [],
+        //     orderTransactionOpts,
+        // );
```

Comment:
Is this outstanding?

Reply:
Yeah. Not sure why it doesnt work. It looks like the Coordinator contract should just be forwarding the transaction to the Exchange contract. Cancellations skip approval verification.

Reply:
It must go through the Coordinator correct? Otherwise the `cancelOrdersUpTo` will cancel orders for tuple
`(makerAddress, NULL_ADDRESS) with salt > targetOrderEpoch`. Rather than `(makerAddress, coordinatorAddress) with salt > targetOrderEpoch `

Reply:
Ahhh ok, I see what I was doing wrong. the test was retrieving orders for  (makerAddress, NULL_ADDRESS) instead of (makerAddress, coordinatorAddress). fixed it :)